RDoc-2499_Corax-6.0-updates

reebhub opened the pull request:
Update the Corax documentation for RavenDB 6.0

maciejaszyk commented on Documentation/6.0/Raven.Documentation.Pages/indexes/boosting.dotnet.markdown:
```diff
@@ -0,0 +1,39 @@
+{INFO Boosting is also available at the query level. You can read more about it [here](../client-api/session/querying/text-search/boost-search-results). /}
+
+{NOTE: }
+When using [Corax](../indexes/search-engine/corax) as the search engine, [boosting is available](../indexes/search-engine/corax#supported-features) 
```
Maybe `indexing-time boosting`, because it may suggest we do not support query-time boosting

reebhub replied:
done

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,377 @@
+{NOTE: }
+
+* **Corax** is RavenDB's native search engine, introduced in RavenDB 
+  version 5.4 as an in-house searching alternative for Lucene.  
```
Maybe "introduced in RavenDB  version 6.0" since in 5.4 it was an experimental feature with very limited set of features it supports. I'd like to avoid confusion that our users think that they can use Corax in 5.4. Corax is 6.0 is completely rewritten feature.

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,377 @@
+{CODE-BLOCK/}
+
+{NOTE: }
+You must restart the server for the new settings to be read and applied.  
```
What might be important info is that an already created index won't change the search engine if you modify the configuration. For example, my current setting is `Indexing.Static.SearchEngineType: Corax`. I have created a couple of static indexes - all of them use Corax. I change the configuration to `Indexing.Static.SearchEngineType: Lucene`. All existing indexes will still use Corax (unless I reset them). Only new indexes are going to use Lucene.

reebhub replied:
{NOTE: }
Selecting a new search engine will change the search engine only for indexes created from now on.  

E.g., If my configuration has been `"Indexing.Static.SearchEngineType": "Corax"` 
until now and i now change it to `"Indexing.Static.SearchEngineType": "Lucene"`, 
static indexes created from now on will use Lucene, but static indexes created 
while Corax was selected will continue using Corax.  

After selecting a new search engine using the above options, change the search 
engine used by an existing index by [resetting](../../client-api/operations/maintenance/indexes/reset-index) the index.  
{NOTE/}

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,377 @@
+
+    | Feature        | Supported by Corax | Comment
+    |----------------|--------------------|--------
+    | Dynamic Fields | `yes` | Corax' handling of dynamic fields is similar to Lucene's.
```
Why we have this comment? What does it mean? What was the intention here? CC @maciejaszyk

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,377 @@
+
+Trying to use Corax with an unimplemented method (see 
+[Supported Features](../../indexes/search-engine/corax#supported-features) above) 
+will generate a `System.NotImplementedException` exception and end the search.  
```
It will throw `NotSupportedInCoraxException`. Right @maciejaszyk ?

reebhub replied:
done

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,377 @@
+
+{PANEL: Handling of Complex JSON Objects}
+
+Complex JSON properties **cannot currently be indexed and searched by Corax**.  
```
Shouldn't it be mentioned in the above section about unsupported features? Maybe with a link in the comment to this section with explains details.

reebhub replied:
right
done

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,377 @@
+You can use Corax as your search engine, but explicitly disable the indexing 
+of complex objects.  
+When you disable the **indexing** of a field this way, the field's contents 
+can still be **stored and projected**.  
```
Maybe rephrase to "can still be stored so it might be used in [projection queries](https://ravendb.net/docs/article-page/5.4/csharp/indexes/querying/projections#projections-and-stored-fields)" with a link to: https://ravendb.net/docs/article-page/5.4/csharp/indexes/querying/projections#projections-and-stored-fields ?

reebhub replied:
done

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,377 @@
+#### If Corax Encounters a Complex Property While Indexing:  
+
+* If an auto index exists for the document, Corax will throw 
+  `System.NotSupportedException` to notify the user that a search 
```
@maciejaszyk I checked the code and indeed it throws `NotSupportedException`. Shouldn't it be changed to NotSupportedInCoraxException?

redknightlois replied:
Yes.

reebhub replied:
@arekpalinski @maciejaszyk pls update me when it's implemented and i'll change it here

arekpalinski replied:
@reebhub this got changed to `NotSupportedInCoraxException`

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,377 @@
+    | | [score()](../../indexes/querying/sorting#ordering-by-score) | `yes` |
+    | | [spatial.distance()](../../client-api/session/querying/how-to-make-a-spatial-query#spatial-sorting) | `yes` |
+
+---
```
Other unsupported features / cases that I found in code.

Querying:

- throw new NotSupportedInCoraxException($"{nameof(Corax)} doesn't support {nameof(Explanations)} yet.");
- throw new NotSupportedInCoraxException($"Corax doesn't support 'Distinct' operation on collection bigger than int32 ({int.MaxValue}).");


Indexing:
- throw new NotSupportedInCoraxException($"{nameof(Corax)} does not support indexing objects that are not points on a world map.");

CC @maciejaszyk

arekpalinski replied:
Also "Custom analyzers" aren't supported in Indexing - https://ravendb.net/docs/article-page/5.4/csharp/studio/database/settings/custom-analyzers

reebhub replied:
added `Explanations`, `Distinct` and **Custom analyzers**, waiting for an answer about _indexing objects_

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,377 @@
+
+* [Indexing.Corax.IncludeSpatialDistance](../../server/configuration/indexing-configuration#indexing.corax.includespatialdistance)
+  Used to include spatial information in document metadata when sorting by distance.  
+
```
We have more Corax specific settings. Look for "Indexing.Corax." here: https://github.com/ravendb/ravendb/blob/v6.0/src/Raven.Server/Config/Categories/IndexingConfiguration.cs

reebhub replied:
done

arekpalinski reviewed (commented):
There are some Corax issues marked with Documentation required tag:
https://issues.hibernatingrhinos.com/issues/RavenDB?q=%23Corax%20AND%20%20%23%7BDocumentation%20Required%7D

@reebhub @maciejaszyk please verify if something needs to be added to the docs. If it's already done then @reebhub please remove the tag.

reebhub commented:
"There are some Corax issues marked with Documentation required tag:
https://issues.hibernatingrhinos.com/issues/RavenDB?q=%23Corax%20AND%20%20%23%7BDocumentation%20Required%7D

@reebhub @maciejaszyk please verify if something needs to be added to the docs. If it's already done then @reebhub please remove the tag."

---

these are the issues, 3 of the 4 are done, the fourth isn't ready for documentation yet.

- [RavenDB-19603](https://issues.hibernatingrhinos.com/issue/RavenDB-19603/Implement-scoring-generated-by-relevance-of-the-documents-boosting) Implement scoring generated by relevance of the documents (boosting) 
	Boosting ranking: added a note to the boosting article
- [RavenDB-19990](https://issues.hibernatingrhinos.com/issue/RavenDB-19990/Allow-to-skip-over-int.Max-docs-in-query-for-Corax-but-limit-projection-up-to-int.Max) Allow to skip over int.Max docs in query (for Corax) but limit projection up to int.Max 
	Added a “limits” section to the corax page and an info box in the paging page.
- [RavenDB-20481](https://issues.hibernatingrhinos.com/issue/RavenDB-20481/Failing-test-on-Corax-Oregon.Fails) Failing test on Corax: Oregon.Fails 
	Added a note to the “Turn the complex property into a string” section.
- [RavenDB-21021](https://issues.hibernatingrhinos.com/issue/RavenDB-21021/Include-Corax-DebugView-expose-query-structure-in-Timings) Include Corax DebugView (expose query structure) in Timings() 
	To be documented when the feature is done.

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,516 @@
+  Set the Corax index compression max documents limit used for dictionary creation.  
+
+* [Indexing.Corax.MaxMemoizationSizeInMb](../../server/configuration/indexing-configuration#indexing.corax.maxmemoizationsizeinmb)
+  The maximum amount of memory that Corax can use for a memoization clause during query processing.  
```
We just added a new config option related to initial dictionary creation:

https://github.com/ravendb/ravendb/pull/17146/files#diff-a26823a405dc6540477281200849f5d7918be61c063cd3228b6034fbbc852fd2R505-R510

reebhub replied:
done

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,516 @@
+* The benefits of compression dictionaries are most pronounced for large collections.  
+* If upon creation there is less than 10000 documents in the collections involved, 
+  it may make sense to manually force an index reset after reaching 100000 documents 
+  to force a retraining.  
```
Note that 10000 docs is configurable via `Indexing.Corax.DocumentsLimitForCompressionDictionaryCreation` mentioned above. Also now we have `Indexing.Corax.MaxAllocationsAtDictionaryTrainingInMb` - 2GB on x64 machines and 128MB on 32bits machines

reebhub replied:
done

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,516 @@
+make the index more readable. 
+
+* **Adding a Compound Field**  
+  In an index definition, add a compound field using the `CompoundFields.Add` method.  
```
This is still in PR but we'll have a strongly types API:

https://github.com/ravendb/ravendb/pull/17176/commits/04d16c6516dd9974c952589f8d049b5907c90e0d

More specifically see the usage here:
https://github.com/ravendb/ravendb/pull/17176/commits/04d16c6516dd9974c952589f8d049b5907c90e0d#diff-948bc0c50471bb01c99e3fac74d035edc0a452dd6015dfad197f36dd03ecfc6dR149-R150

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,516 @@
+
+{PANEL: Compound Fields}
+
+A compound field is a Corax index field comprised of multiple simple data elements.  
```
Currently a compound field can be composed of exactly 2 fields.

reebhub replied:
done

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,516 @@
+# Search Engine: Corax
```
@redknightlois @ayende Can you please review this PR? The most important is this file

redknightlois commented on Documentation/6.0/Raven.Documentation.Pages/server/configuration/indexing-configuration.markdown:
```diff
@@ -418,22 +418,34 @@ Disabling this option can improve query performance.
+- **Default**: `128`
+- **Scope**: Server-wide, per database, or per index
+
+{PANEL/}
```
Mark this as expert level.

reebhub replied:
done

redknightlois commented:
LGTM

redknightlois commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,534 @@
+* If an auto index exists for the document, Corax will throw 
+  `System.NotSupportedException` to notify the user that a search 
+  that makes no sense has been attempted.  
+
```
that in not supported in Corax

reebhub replied:
done

ayende commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+select new
+{
+    // Handling the field as a string will allow Corax to index it
+    Location = order.ShipTo.Location.ToString()
```
That would look the the user as if it would invoke the `Location.ToString()` here, no?

reebhub replied:
--> 
Location = JsonConvert.Serialize(order.ShipTo.Location)

ayende commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+{NOTE/}
+
+{NOTE: }
+Make sure your code handles failure scenarios such as `ToString` returning `null` 
```
This is running server side, so there is not `null` possible, unless I missed something

ayende commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+            select new { p.Brand, p.Location };
+
+        // Add a compound field 
+        CompoundField(x => x.Brand, x.Location);
```
```suggestion
        CompoundField(x => x.Brand, x => x.Location);
```

reebhub replied:
done

ayende commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+{CODE-BLOCK/}
+
+    The query that uses the indexed data will look no different than if the 
+    index included no compound field, but produce the results much faster.  
```
Should add a note that this should be applied on very large datasets and specific queries. This is very much an expert only option.

reebhub replied:
added

ayende commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+
+{PANEL: Limits}
+
+* Corax can create and use indexes of more than `int.MaxValue` documents.  
```
`int.MaxValue` to number

ayende commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+* [Indexing.Static.SearchEngineType](../../server/configuration/indexing-configuration#indexing.static.searchenginetype)  
+  [Select](../../indexes/search-engine/corax#selecting-the-search-engine) the search engine for **Static** indexes.  
+
+* [Indexing.Corax.IncludeDocumentScore](../../server/configuration/indexing-configuration#indexing.corax.includedocumentscore)  
+  Choose whether to include the score value in document metadata when sorting by score.  
+
+* [Indexing.Corax.IncludeSpatialDistance](../../server/configuration/indexing-configuration#indexing.corax.includespatialdistance)  
+  Choose whether to include spatial information in document metadata when sorting by distance.  
```
Why would I want one or the other? 
I assume there is a perf cost associated with enabling thos?

reebhub replied:
@redknightlois

arekpalinski replied:
Yes, IIRC this is related to performance. CC @maciejaszyk

ayende commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+  {NOTE: }
+  Training stops when it reaches either the 
+  [number of documents](../../server/configuration/indexing-configuration#indexing.corax.documentslimitforcompressiondictionarycreation)
+  threshold (10000 docs by default) or the 
```
```suggestion
  threshold (10,000 docs by default) or the 
```

reebhub replied:
done

ayende commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+  [number of documents](../../server/configuration/indexing-configuration#indexing.corax.documentslimitforcompressiondictionarycreation)
+  threshold (10000 docs by default) or the 
+  [amount of memory](../../server/configuration/indexing-configuration#indexing.corax.maxallocationsatdictionarytraininginmb) 
+  threshold (2GB/128MB by default). Both thresholds are configurable.  
```
What is 128MB here? You refer to 32 bits, which can safely be skipped

reebhub replied:
removed 128MB

ayende commented on Documentation/6.0/Raven.Documentation.Pages/server/configuration/indexing-configuration.markdown:
```diff
@@ -0,0 +1,471 @@
+Training will stop when it reaches this limit.  
+
+- **Type**: `SizeUnit.Megabytes`
+- **Default**: 2 GB for x64, or 128 MB for x86 (32 bits)
```
```suggestion
- **Default**: 2 GB for 64 bits systems, or 128 MB for 32 bits systems
```

reebhub replied:
done

ayende commented on Documentation/6.0/Raven.Documentation.Pages/server/configuration/indexing-configuration.markdown:
```diff
@@ -0,0 +1,471 @@
+Training will stop when it reaches this limit.  
+
+- **Type**: `int`
+- **Default**: `100000`
```
That says 10K above, which one is right?

reebhub replied:
right
fixed

Danielle9897 commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+
+    | Feature                                                     | Supported by Corax 
+    |------------------------------------------------------------------------|--------------------
+    | [MoreLikeThis](../../indexes/querying/morelikethis) | `yes` 
```
Add that **Custom Sorters** are Not supported by Corax as well

reebhub replied:
added, thanks

reebhub commented on Documentation/6.0/Raven.Documentation.Pages/indexes/querying/paging.dotnet.markdown:
```diff
@@ -11,6 +11,14 @@ By default, if the number of returned results exceeds **2048**, the server will
 
+{INFO:Limits}
+When [Corax](../../indexes/search-engine/corax) is used as the search engine, 
+indexes of more than `int.MaxValue` documents can be created and used.  
```
done

reebhub commented on Documentation/6.0/Raven.Documentation.Pages/indexes/querying/paging.java.markdown:
```diff
@@ -11,6 +11,14 @@ By default, if the number of returned results exceeds **2048**, the server will
 
+{INFO:Limits}
+When [Corax](../../indexes/search-engine/corax) is used as the search engine, 
+indexes of more than `int.MaxValue` documents can be created and used.  
```
done

reebhub commented on Documentation/6.0/Raven.Documentation.Pages/indexes/querying/paging.js.markdown:
```diff
@@ -11,6 +11,14 @@ By default, if the number of returned results exceeds **2048**, the server will
 
+{INFO:Limits}
+When [Corax](../../indexes/search-engine/corax) is used as the search engine, 
+indexes of more than `int.MaxValue` documents can be created and used.  
```
done

reebhub commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+
+* You can select your preferred search engine in several scopes:  
+   * [Server-wide](../../indexes/search-engine/corax#select-search-engine-server-wide), 
+     selecting which search engine will be used by all the databases hosted by this server.  
```
done

reebhub commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+    | Searching by [Regex](../../client-api/session/querying/text-search/using-regex) | `yes` 
+    | [Fuzzy Search](../../client-api/session/querying/text-search/fuzzy-search) | **no** 
+    | [Explanations](../../client-api/session/querying/debugging/include-explanations) | **no** 
+    | [Distinct](../../indexes/querying/distinct) operation on a collection with more than int32 ({int.MaxValue}) documents | **no** 
```
which is better, removing it completely or stating it isn't supported by both?

reebhub commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+    | [Explanations](../../client-api/session/querying/debugging/include-explanations) | **no** 
+    | [Distinct](../../indexes/querying/distinct) operation on a collection with more than int32 ({int.MaxValue}) documents | **no** 
+
+* **Dynamic Fields**  
```
done

reebhub commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,538 @@
+
+As the `Location` property of the document above contains a list of key/value pairs 
+rather than a simple numeric value or a string, attempting to index this field using 
+Corax [would fail](../../indexes/search-engine/corax#if-corax-encounters-a-complex-property-while-indexing).  
```
{NOTE: }
The approach taken by Lucene, indexing such objects as a JSON string, usually makes no sense and is not supported by Corax.  
{NOTE/}

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,501 @@
+select new
+{
+    // Handling the field as a string will allow Corax to index it
+    Location = JsonConvert.Serialize(order.ShipTo.Location)
```
Such index will not compile:

```
(25,32): error CS0103: The name 'JsonConvert' does not exist in the current context
, IndexDefinitionProperty='', ProblematicText=''   at Raven.Server.Documents.Indexes.Static.IndexCompiler.CompileInternal(String originalName, String cSharpSafeName, MemberDeclarationSyntax class, IndexDefinition definition) in C:\Builds\RavenDB-6.0-Nightly\20230907-0851\src\Raven.Server\Documents\Indexes\Static\IndexCompiler.cs:line 244
```

reebhub replied:
done

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,507 @@
+  [number of documents](../../server/configuration/indexing-configuration#indexing.corax.documentslimitforcompressiondictionarycreation)
+  threshold (100,000 docs by default) or the 
+  [amount of memory](../../server/configuration/indexing-configuration#indexing.corax.maxallocationsatdictionarytraininginmb) 
+  threshold (2GB by default). Both thresholds are configurable.  
```
Update: We changed that so now it's dependent the the amount of installed memory on the machine:
https://github.com/ravendb/ravendb/pull/17298/files#diff-a26823a405dc6540477281200849f5d7918be61c063cd3228b6034fbbc852fd2R44-R72

Maybe just mention that without saying explicitly exact number. It's up to 2GB.

reebhub replied:
--> or the [amount of memory](../../server/configuration/indexing-configuration#indexing.corax.maxallocationsatdictionarytraininginmb) threshold (up to 2GB).

changed in https://github.com/ravendb/docs/pull/1695

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,507 @@
+Corax indexes will not train compression dictionaries if they are created in the 
+testing studio interface, because it is designed for indexing prototyping and the 
+training process will add unnecessary overhead.
+{NOTE/}
```
Here we could reference Test Index feature article (once we'll have it)

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,507 @@
+    ![Index Definition](images/corax-02_index-definition.png "Index Definition")
+    1. Open the index **Configuration** tab.  
+    2. Select the search engine you prefer for this index.  
+       ![Per-Index Search Engine](images/corax-03_index-definition_searcher-select.png "Per-Index Search Engine")
```
I see on this image `Corax - experimental` - it's no longer experimental. We need to update the image.

reebhub replied:
retook all snapshots for the article

arekpalinski commented on Documentation/6.0/Raven.Documentation.Pages/indexes/search-engine/corax.markdown:
```diff
@@ -0,0 +1,507 @@
+#### If Corax Encounters a Complex Property While Indexing:  
+
+* If an auto index exists for the document, Corax will throw 
+  `System.NotSupportedException` to notify the user that a search 
```
Update: We're gonna change that behavior for auto indexes: https://issues.hibernatingrhinos.com/issue/RavenDB-21430/Corax-Alert-users-about-indexing-of-complex-objects-in-auto-indexes-instead-of-throwing-an-error

reebhub replied:
--> 

* If an auto index exists for the document Corax will alert the user:  
  `The value of '{fieldName}' field is a complex object. Indexing it as a text isn't supported. You should consider querying on individual fields of that object.`

changed in https://github.com/ravendb/docs/pull/1695
